### ISSUE TITLE
beta-channel hooks

### ISSUE THREAD

**Issue by EliRibble** [pre-commit/pre-commit#1340]
I manage the pre-commit hooks for a large organization of engineers. I'd like to set up a way for a subset of those engineers to test out new hooks on an experimental basis. Ideally I'd be able to tag or categorize some set of the pre-commit hooks as stable and another set as 'beta' or 'unstable' or something. Engineers could then opt-in to the beta/unstable channel and the hooks would run for their changes but there would be an easy way for them, on an individual-commit basis, to skip those beta checks when they are wrong.

I couldn't figure out a way to do this without making engineers be very involved with using `SKIP` env vars with pre-commit as it presently stands.

I'd be interested in writing this functionality myself, but it'd be good to get agreement on exactly how it should behave. Here's my proposal.

1. Add a new configuration option to `.pre-commit-config.yaml` at  `hooks[].tags`. This is a list of strings that can apply to any single hook in `.pre-commit-config.yaml`.
2. Add a new configuration option to `.pre-commit-config.yaml` at `hooks[].blocking`. This value defaults to `True`. When set to `False` the hook can fail and will still print the failure output but not prevent the commit.
3. Update the output of the report from `pre-commit` so it includes the tags next to the name.
4. Allow `SKIP` to match both the hook name and one or more of a hooks tags.
5. Create a new environment variable, `PRECOMMIT_BLOCKING`, which is a comma-separated list of tags/hook names to make blocking, overridding the configuration file's new `hooks[].blocking` setting.

Combined together this allows a developer to add a new pre-commit hook and tag it as `beta/unstable` and mark it unblocking by default. Then a small team working on the tag can mark the tag(s) as blocking for them via `PRECOMMIT_BLOCKING` while they refine the hooks and deal with feedback from the larger team. After (months) of testing they can eventually switch `blocking` so the rest of the team has to pay attention to the new hook.

Would the maintainers be interested in a pull request with this functionality? If not, what changes would you like to see?

**Comment by asottile** [comment]
I'm not really interested in doing any of the things as you've described -- sorry.  just sounds like functionality that's completely out of scope, difficult to maintain, and with dubious-at-best goals

How about setting up `.pre-commit-config.beta.yaml` with the alternate versions. individauls can choose to `pre-commit install --config .pre-commit-config.beta.yaml` and utilize the new versions without impacting the critical path

It should be pretty easy / sufficient for you to run `pre-commit try-repo --all-files` as you develop hooks (or work on higher test coverage in the tools, etc.)

**Comment by EliRibble** [comment]
No need to feel sorry - I haven't been able to find a clear delineation of what's considered in-scope for this project, so asserting it clearly is helpful!

Admittedly I likely know less about this than you, but I generally get complaints from `pre-commit` when I've made local changes to the `.pre-commit-config.yaml` file without having committed it. In your proposed workflow with `.pre-commit-config.beta.yaml` would `pre-commit` complain if I've got changes to either `.pre-commit-config.beta.yaml` or `.pre-commit-config.yaml` when I have installed `.pre-commit-config.beta.yaml`?

If not, I think that solution would work fine. I'm mostly just looking for a way to have a subset of engineers have an experimental configuration they opt into without having to mess up everyone else who does not develop on the hooks, but does rely on them.

**Comment by asottile** [comment]
I believe the unstaged check only applies to the configuration pre-commit is sourcing from (I'd have to try it to completely verify -- but at least that's the intent) -- the reason such a check exists is pre-commit "stashes" the unstaged changes such that it is validating the code that is actually being committed (instead of whatever happens to be in your working directory -- a lot of other frameworks get this _very wrong_ and allow you to commit invalid code because of it).  The reason the configuration itself is particularly sensitive is pre-commit has to read that from disk (this produces a bit of confusion in both directions -- either you commit against the wrong set of hooks or pre-commit runs against something isn't what's reflected on disk) -- as such pre-commit enforces that this file is always cleanly staged or unedited.

but yeah, off the top of my head not sure of the behaviour here -- if it requires both files to be unstaged then there's a bug that should get fixed 👍

**Comment by asottile** [comment]
@EliRibble did that work for you?

**Comment by EliRibble** [comment]
Sorry, I haven't tested it yet, I've been on vacation. I should be back and
try it in 1 week.

**Comment by EliRibble** [comment]
Alright, I've tested this now. It works in the sense that I can install the .pre-commit-config.beta.yaml file and it will correctly run a different set of hooks than what is defined in .pre-commit-config.yaml. Unfortunately it looks like things like `pre-commit run ...` will still use the default config unless I specify `pre-commit --config .pre-commit-config.beta.yaml run ...`.

That said, I think it'll be good enough for my use case for now, thank you!

**Comment by asottile** [comment]
🎉 glad that works for ya -- and thanks for the issue!
